### PR TITLE
Replaced BiRewire with bipartite edge swap

### PR DESCRIPTION
**Pull request by matthewreyna**
This pull request replaces BiRewire with a bipartite edge swapping algorithm for drawing from the permutational distribution.  I implemented the bipartite edge swapping algorithm in Fortran, and we can import it as a Python module.

The ECCB experiments now complete faster with much less file system use, and the results are similar: I found 48, 5341, and 6807 gene set triples with a weighted FDR < 0.001 for the THCA, COADREAD, and UCEC datasets, respectively, while using only 1000 permutations for testing.

### CODE REVIEW

File: compute_mutation_probabilities.py [hunk 1]
```diff
@@ -2,97 +2,140 @@
...
         map_fn = pool.map
     else:
         map_fn = map
 
-    compute_args = [ (permuted_files[i::args.num_cores], geneToIndex, patientToIndex)
-                     for i in range(args.num_cores) ]
-
-    observeds = map_fn(compute_observed_wrapper, compute_args)
+    wrapper_args = [ (edge_list, max_swaps, max_tries, seeds[i::10*num_cores], 0, m,
+                      n, num_edges, indexToGene, indexToPatient) for i in range(10*num_cores) ]
```
Comment (mdml) [inline compute_mutation_probabilities.py]: Why is this `10*num_cores`?

File: compute_mutation_probabilities.py [hunk 2]
```diff
@@ -2,97 +2,140 @@
...
+            assert( np.abs(P[:, patientToIndex[p]-1].sum() - obs) < 0.1)
+
+        # Add pseudocounts to entries with no mutations observed
+        P[P == 0] = 1./(2. * args.num_permutations)
+
+        # Output to file.
+        # The rows/columns preserve the order given by the mutation file.
+        np.save(args.weights_file, P)
 
-    # Verify the weights
```
Comment (mdml) [inline compute_mutation_probabilities.py:84]: Why delete verification?

File: experiments/eccb2016/commands.sh [hunk 3]
```diff
@@ -9,8 +9,8 @@ THCA_MIN_FREQ=5
 UCEC_MIN_FREQ=30
 LENGTH_THRESHOLD=600
 FDR_CUTOFF=0.001
-TOTAL_PERMUTATIONS=10000 # set to 1000000 for ECCB 2016 submission
-PAIRS_PERMUTATIONS=10000
+TOTAL_PERMUTATIONS=1000 # set to 1000000 for ECCB 2016 submission
+PAIRS_PERMUTATIONS=1000
```
Comment (mdml) [inline experiments/eccb2016/commands.sh]: This should be 10000.

File: experiments/eccb2016/commands.sh [hunk 4]
```diff
@@ -255,64 +181,23 @@ python compute_exclusivity.py -mf $UCEC_MUTATIONS -nc $NUM_CORES -k 3 \
...
-       -uf $THCA_UNWEIGHTED_EXACT_TRIPLES -nt 5 -lt $LENGTH_THRESHOLD \
+       -uf $THCA_UNWEIGHTED_SADDLEPOINT_TRIPLES -nt 5 -lt $LENGTH_THRESHOLD \
        -f $FDR_CUTOFF -o $TABLES_DIR/Table4
 
 # TABLE 5: COADREAD results
 echo "COADREAD"
 python $SCRIPTS_DIR/results_table.py -lf $GENE_LENGTH_FILE \
        -mf $COADREAD_MUTATIONS -wf $COADREAD_WEIGHTED_SADDLEPOINT_TRIPLES \
-       -uf $COADREAD_UNWEIGHTED_EXACT_TRIPLES -nt 5 -lt $LENGTH_THRESHOLD \
+       -uf $COADREAD_UNWEIGHTED_SADDLEPOINT_TRIPLES -nt 5 -lt $LENGTH_THRESHOLD \
```
Comment (mdml) [inline experiments/eccb2016/commands.sh]: Why?

File: compute_mutation_probabilities.py (hunk 2 above)
Comment (matthewreyna) [inline compute_mutation_probabilities.py:84]: Unintentional.

File: compute_mutation_probabilities.py (hunk 1 above)
Comment (matthewreyna) [inline compute_mutation_probabilities.py]: We should not bunch our runs into `num_cores` chunks.  The `chunksize` argument for `map` is a better way to do this, but not a priority.

File: experiments/eccb2016/commands.sh (hunk 3 above)
Comment (matthewreyna) [inline experiments/eccb2016/commands.sh]: Fixed.

File: experiments/eccb2016/commands.sh (hunk 4 above)
Comment (matthewreyna) [inline experiments/eccb2016/commands.sh]: I am no longer computing the exact results here.  We can run an earlier, tagged version of `commands.sh` to reproduce the full results from the paper.

Comment (mdml) [inline experiments/eccb2016/commands.sh]: I'm not sure this is a good idea. It's one thing if `experiments/eccb2016/commands.sh` doesn't run unless you are at a certain tagged commit. It's another if it runs and produces something different, don't you think?

**Comment by mdml** [comment]
LGTM 👍

**Pull request merged**